File: Stage30/Java/Problem1450.java
package Stage30.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Problem1450 {
    static int N, C;
    static int[] arr;
    static ArrayList<Integer> left = new ArrayList<>();
    static ArrayList<Integer> right = new ArrayList<>();
    
    static void dfs(int start, int end, int sum, ArrayList<Integer> list) {
        if (sum > C) return;

        if (start == end) {
            list.add(sum);
            return;
        }

        dfs(start + 1, end, sum, list);

        dfs(start + 1, end, sum + arr[start], list);
    }

    static int upperBound(ArrayList<Integer> list, int target) {
        int left = 0;
        int right = list.size();

        while (left < right) {
            int mid = (left + right) / 2;
            if (list.get(mid) <= target) left = mid + 1;
            else right = mid;
        }

        return left;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        dfs(0, N / 2, 0, left);
        dfs(N / 2, N, 0, right);

        Collections.sort(right);

        long count = 0;

        for (int l : left) {
            if (l > C) continue;

            int remain = C - l;

            int idx = upperBound(right, remain);
            count += idx;
        }

        System.out.println(count);
        br.close();
    }
}

File: Stage30/C/Problem1450.c
#include <stdio.h>
#include <stdlib.h>

int N, C;
int arr[30];

int left[1 << 15];
int right[1 << 15];
int l_size = 0, r_size = 0;

void dfs(int start, int end, int sum, int* list, int* size) {
    if (sum > C) return;

    if (start == end) {
        list[(*size)++] = sum;
        return;
    }

    dfs(start + 1, end, sum, list, size);

    dfs(start + 1, end, sum + arr[start], list, size);
}

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int upper_bound(int* arr, int size, int target) {
    int left = 0, right = size;

    while (left < right) {
        int mid = (left + right) / 2;
        if (arr[mid] <= target) left = mid + 1;
        else right = mid;
    }

    return left;
}

int main(void) {
    scanf("%d %d", &N, &C);

    for (int i = 0; i < N; i++) {
        scanf("%d", &arr[i]);
    }

    dfs(0, N / 2, 0, left, &l_size);
    dfs(N / 2, N, 0, right, &r_size);

    qsort(right, r_size, sizeof(int), compare);

    long long count = 0;

    for (int i = 0; i < l_size; i++) {
        if (left[i] > C) continue;

        int remain = C - left[i];

        int idx = upper_bound(right, r_size, remain);
        count += idx;
    }

    printf("%lld\n", count);

    return 0;
}